pull or fetch available?

maybe a duplicate of #5 (?) 
or maybe I have overlooked something, 
but I did not find a way to `pull` l or `fetch` in gittyup or hlibgit2.
does it exist, or are there plans to support it? 
or is it blocked somewhere upstream? 

thank you. 🙂

```nim
import os
import nimgit2
import strformat
import strutils

type
  Repo = git_repository
  GitOId = git_oid
  GitTime = git_time_t
  StrArray = git_strarray
  GitReference = git_reference
  Commit = git_commit

func `$`*(oid: var GitOId): string =
  result = $git_oid_tostr_s(addr oid)

proc info(text: string) {.discardable.} =
  echo "[INFO] ", text

proc error(text: string) {.discardable.} =
  echo "[ERROR] ", text

proc reportResult(resultCode: int, resultName: string) {.discardable.} =
  if resultCode < 0:
    let lastErr = git_error_last()
    ($(lastErr.klass) & ": " & $(lastErr.message)).error
  else:
    # "success: {resultName}: {resultCode}".info
    return

proc getOidFromSha(sha: cstring): GitOId =
  reportResult(git_oid_fromstr(addr result, sha), "getOidFromSha")

proc getShaFromOid(oid: var GitOId): cstring =
  return git_oid_tostr_s(addr oid)

proc getCommitFromSha(repo: ptr Repo, sha: cstring): ptr Commit =
  var
    shaOid = getOidFromSha(sha)
    shaCommit: ptr git_commit
  let commitFromShaResult = git_commit_lookup(addr shaCommit, repo, addr shaOid)
  reportResult(commitFromShaResult, "commitFromShaResult")
  return shaCommit

proc report_merge_analysis(result: git_merge_analysis_t) =
  if   result == GIT_MERGE_ANALYSIS_NONE:        fmt"GIT_MERGE_ANALYSIS_NONE       : {GIT_MERGE_ANALYSIS_NONE}"       .info
  elif result == GIT_MERGE_ANALYSIS_NORMAL:      fmt"GIT_MERGE_ANALYSIS_NORMAL     : {GIT_MERGE_ANALYSIS_NORMAL}"     .info
  elif result == GIT_MERGE_ANALYSIS_UP_TO_DATE:  fmt"GIT_MERGE_ANALYSIS_UP_TO_DATE : {GIT_MERGE_ANALYSIS_UP_TO_DATE}" .info
  elif result == GIT_MERGE_ANALYSIS_FASTFORWARD: fmt"GIT_MERGE_ANALYSIS_FASTFORWARD: {GIT_MERGE_ANALYSIS_FASTFORWARD}".info
  elif result == GIT_MERGE_ANALYSIS_UNBORN:      fmt"GIT_MERGE_ANALYSIS_UNBORN     : {GIT_MERGE_ANALYSIS_NONE}"       .info
  fmt"GIT_MERGE_ANALYSIS: {result}".info

const
  repoPath = "/tmp/gittyup"
  branchName = "master"
  remoteName = "origin"
  url = "https://github.com/disruptek/gittyup"
  masterRef = "refs/remotes/origin/master"
  logMessage = "attempt to fetch / pull fast forward"
  refLogMessage = "fetch"
  theirHeadsLen = 1.cuint

let initResult = git_libgit2_init()
reportResult(initResult, "init_result")

var
  repo: ptr Repo
  repoHeadRef: ptr GitReference
  fetchOpts: git_fetch_options
  remote: ptr git_remote
  refSpecs: StrArray
  gitRef: ptr GitReference
  fetchAnnoCommit: ptr git_annotated_commit
  checkoutOpt: git_checkout_options
  mergeAnalysis: git_merge_analysis_t
  mergePreference: git_merge_preference_t
  newTargetRef: ptr GitReference

fmt"clone or pull of {url} at {repoPath}".info
if not repoPath.dir_exists():
  "repo does not exist yet: cloning..".info
  let cloneOpenResult = git_clone(addr repo, url, repoPath, nil)
  reportResult(cloneOpenResult, "cloneOpenResult")
else:
  "repo exists: opening..".info
  let openResult = git_repository_open(addr repo, repoPath)
  reportResult(openResult, "openResult")

let repoHeadRefResult = git_repository_head(addr repoHeadRef, repo)
reportResult(repoHeadRefResult, "repoHeadRefResult")

"fetching..".info
let optInitResult = git_fetch_options_init(addr fetchOpts, 1.cuint)
reportResult(optInitResult, "optInitResult")

let remoteResult = git_remote_lookup(addr remote, repo, remoteName)
reportResult(remoteResult, "remoteResult")

let fetchResult = git_remote_fetch(remote, addr refSpecs, addr fetchOpts, refLogMessage)
reportResult(fetchResult, "fetchResult")

let remoteMasterIdResult = git_reference_lookup(addr gitRef, repo, masterRef)
reportResult(remoteMasterIdResult, "remoteMasterIdResult")
let refTarget = git_reference_target(git_ref)

let fetchAnnoRes = git_annotated_commit_from_fetchhead(addr fetchAnnoCommit,
    repo, branchName, url, refTarget)
reportResult(fetchAnnoRes, "fetchAnnoRes")

let analyseResult = git_merge_analysis(addr mergeAnalysis, addr mergePreference,
    repo, addr fetchAnnoCommit, theirHeadsLen)

reportResult(analyseResult, "analyseResult")
reportMergeAnalysis(mergeAnalysis)

if   mergeAnalysis == GIT_MERGE_ANALYSIS_UP_TO_DATE:
  "nothing to do: UP_TO_DATE".info

elif mergeAnalysis == GIT_MERGE_ANALYSIS_FASTFORWARD:
  "WIP: GIT_MERGE_ANALYSIS_FASTFORWARD".info

elif mergeAnalysis == 5:
  "WIP: GIT_MERGE_ANALYSIS: NORMAL & FASTFORWARD".info

  var optRes = git_checkout_options_init(addr checkoutOpt, GIT_CHECKOUT_OPTIONS_VERSION)
  reportResult(optRes, "optRes")

  let remoteMasterId = getShaFromOid(ref_target[])
  fmt"{remoteMasterId =}".info

  let
    remoteMasterCommit = getCommitFromSha(repo, remoteMasterId)
    commitToObj = cast[ptr git_object](remoteMasterCommit)
    checkoutResult = git_checkout_tree(repo, commitToObj, addr checkoutOpt)
  reportResult(checkoutResult, "checkoutResult")

  let setTargetRes = git_reference_set_target(addr newTargetRef, repoHeadRef,
      refTarget, logMessage)
  reportResult(setTargetRes, "setTargetRes")

elif mergeAnalysis == GIT_MERGE_ANALYSIS_NORMAL:
  "WIP: GIT_MERGE_ANALYSIS_NORMAL".info

else:
  "WIP: other GIT_MERGE_ANALYSIS".info

let shutdownResult = git_libgit2_shutdown()
fmt"{shutdownResult =}".info
```
^^ this is the snippet @fbpyr mentioned.
I once wrote it for using it as autoupdater for some repos, 
where it still works pretty well.  No tests so far, though.

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -900,16 +900,6 @@ proc remoteLookup*(repo: GitRepository; name: string): GitResult[GitRemote] =
       assert remote != nil
       result.ok remote
 
-
-proc remoteLookup*(repo: GitRepository; name: string): GitResult[GitRemote] =
-  ## get the remote by name; the remote must be freed
-  withGit:
-    var
-      remote: GitRemote
-    withResultOf git_remote_lookup(addr remote, repo, name):
-      assert remote != nil
-      result.ok remote
-
 proc remoteRename*(repo: GitRepository; prior: string;
                    next: string): GitResult[seq[string]] =
   ## rename a remote

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -843,6 +843,21 @@ proc openRepository*(path: string): GitResult[GitRepository]
   ## alias for `repositoryOpen`
   result = repositoryOpen(path)
 
+proc fetch*(repo: GitRepository): GitResultCode =
+  withGit:
+    var
+      fetchOpts: git_fetch_options
+      refSpecs: git_strarray
+      remote: ptr git_remote
+    try:
+      withResultOf git_fetch_options_init(addr fetchOpts, GIT_CLONE_OPTIONS_VERSION):
+        withResultOf git_remote_lookup(addr remote, repo, "origin"):
+          result = git_remote_fetch(remote, addr refSpecs, addr fetchOpts, "fetch").grc
+    finally:
+      git_strarray_dispose(addr refSpecs)
+      dealloc addr fetchOpts
+      dealloc addr remote
+
 proc repositoryHead*(repo: GitRepository): GitResult[GitReference] =
   ## fetch the reference for the repository's head; the reference must be freed
   withGit:

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -885,6 +885,22 @@ proc getRemoteNames*(repo: GitRepository): seq[string] =
     finally:
       git_strarray_dispose(addr remotesList)
 
+proc fetchRemotes*(repo: GitRepository, remoteNames: seq[string]): seq[GitResultCode] =
+  ## fetch from repo at given remoteNames
+  withGit:
+    for remoteName in remoteNames:
+      result.add(fetchRemote(repo, remoteName))
+
+proc remoteLookup*(repo: GitRepository; name: string): GitResult[GitRemote] =
+  ## get the remote by name; the remote must be freed
+  withGit:
+    var
+      remote: GitRemote
+    withResultOf git_remote_lookup(addr remote, repo, name):
+      assert remote != nil
+      result.ok remote
+
+
 proc remoteLookup*(repo: GitRepository; name: string): GitResult[GitRemote] =
   ## get the remote by name; the remote must be freed
   withGit:

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -873,6 +873,18 @@ proc headReference*(repo: GitRepository): GitResult[GitReference] =
   ## alias for repositoryHead
   result = repositoryHead(repo)
 
+proc getRemoteNames*(repo: GitRepository): seq[string] =
+  ## get names of all remotes
+  withGit:
+    var remotesList: git_str_array
+    try:
+      withResultOf git_remote_list(addr remotesList, repo):
+        let remoteNames = cstringArrayToSeq(cast[cstringArray](remotesList.strings), remotesList.count)
+        for name in remoteNames:
+          result.add(name.string)
+    finally:
+      git_strarray_dispose(addr remotesList)
+
 proc remoteLookup*(repo: GitRepository; name: string): GitResult[GitRemote] =
   ## get the remote by name; the remote must be freed
   withGit:

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -844,7 +844,7 @@ proc openRepository*(path: string): GitResult[GitRepository]
   ## alias for `repositoryOpen`
   result = repositoryOpen(path)
 
-proc fetch*(repo: GitRepository, remoteName: string): GitResultCode =
+proc fetchRemote*(repo: GitRepository, remoteName: string): GitResultCode =
   ## fetch from repo at given remoteName
   withGit:
     var

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -45,6 +45,7 @@ const
   GIT_STATUS_OPTIONS_VERSION* = 1
   GIT_CLONE_OPTIONS_VERSION* = 1
   GIT_CHECKOUT_OPTIONS_VERSION* = 1
+  GIT_FETCH_OPTIONS_VERSION* = 1
 
 # git_strarray_dispose replaces git_strarray_free in >v1.0.1
 when not compiles(git_strarray_dispose):
@@ -843,15 +844,16 @@ proc openRepository*(path: string): GitResult[GitRepository]
   ## alias for `repositoryOpen`
   result = repositoryOpen(path)
 
-proc fetch*(repo: GitRepository): GitResultCode =
+proc fetch*(repo: GitRepository, remoteName: string): GitResultCode =
+  ## fetch from repo at given remoteName
   withGit:
     var
       fetchOpts: git_fetch_options
       refSpecs: git_strarray
       remote: ptr git_remote
     try:
-      withResultOf git_fetch_options_init(addr fetchOpts, GIT_CLONE_OPTIONS_VERSION):
-        withResultOf git_remote_lookup(addr remote, repo, "origin"):
+      withResultOf git_fetch_options_init(addr fetchOpts, GIT_FETCH_OPTIONS_VERSION):
+        withResultOf git_remote_lookup(addr remote, repo, remoteName.cstring):
           result = git_remote_fetch(remote, addr refSpecs, addr fetchOpts, "fetch").grc
     finally:
       git_strarray_dispose(addr refSpecs)

--- a/gittyup.nim
+++ b/gittyup.nim
@@ -858,7 +858,7 @@ proc fetchRemote*(repo: GitRepository, remoteName: string): GitResultCode =
     finally:
       git_strarray_dispose(addr refSpecs)
       dealloc addr fetchOpts
-      dealloc addr remote
+      free remote
 
 proc repositoryHead*(repo: GitRepository): GitResult[GitReference] =
   ## fetch the reference for the repository's head; the reference must be freed

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -168,4 +168,4 @@ suite "giddy up, pardner":
     cloned := cloneme.clone(tmpdir):
       fail dumpError(code)
     check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
-    check cloned.fetch("origin") == GIT_OK
+    check cloned.fetchRemote("origin") == GIT_OK

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -170,10 +170,3 @@ suite "giddy up, pardner":
     check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
     check cloned.fetchRemote("origin") == GIT_OK
 
-  ## fetchRemotes
-  test:
-    cloned := cloneme.clone(tmpdir):
-      fail dumpError(code)
-    check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
-    for gitResult in cloned.fetchRemotes(getRemoteNames(cloned)):
-      check gitResult == GIT_OK

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -163,7 +163,7 @@ suite "giddy up, pardner":
         free thing
       fail "unable to find v102"
 
-  ## fetch
+  ## fetchRemote
   test:
     cloned := cloneme.clone(tmpdir):
       fail dumpError(code)

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -162,3 +162,10 @@ suite "giddy up, pardner":
           break found
         free thing
       fail "unable to find v102"
+
+  ## fetch
+  test:
+    cloned := cloneme.clone(tmpdir):
+      fail dumpError(code)
+    check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
+    check cloned.fetch() == GIT_OK

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -168,4 +168,4 @@ suite "giddy up, pardner":
     cloned := cloneme.clone(tmpdir):
       fail dumpError(code)
     check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
-    check cloned.fetch() == GIT_OK
+    check cloned.fetch("origin") == GIT_OK

--- a/tests/test.nim
+++ b/tests/test.nim
@@ -169,3 +169,11 @@ suite "giddy up, pardner":
       fail dumpError(code)
     check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
     check cloned.fetchRemote("origin") == GIT_OK
+
+  ## fetchRemotes
+  test:
+    cloned := cloneme.clone(tmpdir):
+      fail dumpError(code)
+    check GIT_REPOSITORY_STATE_NONE == cloned.repositoryState
+    for gitResult in cloned.fetchRemotes(getRemoteNames(cloned)):
+      check gitResult == GIT_OK

